### PR TITLE
Use one AllReduceParameter for multi-optim method  training

### PR DESCRIPTION
**Pull request by Menooker**
## What changes were proposed in this pull request?

For multiple layers in DirtriOptimizer, use one shared AllReduceParameter instead of different AllReduceParameters for each layer. Use a map parameterSplit to hold the mapping from layer name to (startIdx, length) of the layer. The optimizer will apply its gradient partition on its weight partition for each layer and if the layer's parameterSplit is out of the partition, the optimizer will truncate or skip the gradient to find an appropriate tensor slice for local gradient partition and weight partition.

## How was this patch tested?

UT

## Related links or issues (optional)
Preliminary changes for PR https://github.com/intel-analytics/BigDL/pull/2810

### CODE REVIEW

**Comment by jason-dai** [comment]
Why is there LarsSGD in this PR?

**Comment by Menooker** [comment]
> Why is there LarsSGD in this PR?

The commits of LarsSGD has already been merged to master. It may be an Github issue, since the commit IDs are identical to the merged commits and Github shows this PR has no conflicts. Please ignore the LARS-related changes.

**Comment by Menooker** [comment]
@jason-dai I have manually rebase the commits. The LarsSGD commits should be gone.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 1]
```diff
@@ -135,6 +131,15 @@ class AllReduceParameter[T: ClassTag](
       .getOrElse(throw new IllegalStateException("Please initialize AllReduceParameter first!"))
   }
 
+  def localPartitionRange: (Int, Int) = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: private[bigdl] ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 2]
```diff
@@ -135,6 +131,15 @@ class AllReduceParameter[T: ClassTag](
...
+    val taskSize = size / partitionNum
+    val extraSize = size % partitionNum
+    val partitionId = TaskContext.getPartitionId()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: Are you sure the `partitionId` is the same as the `partitionId` of this object ?  use the global one instead

Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: This method should only be used in Rdd.mapPartitions(...). The partitionId here should be the id from the `mapPartitions`

Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: @wzhongyuan Changed as required. Found that `partitionId` is already valid when calling localPartitionRange

**Comment by wzhongyuan** [comment]
Please run a Jenkins pass after verify the compatibility

**Comment by Menooker** [comment]
jenkins passed: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/602/

**Comment by wzhongyuan** [comment]
LGTM

**Comment by Menooker** [comment]
Jenkins passed http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/609/

**Comment by Menooker** [comment]
Jenkins fixed: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/615/

**Pull request merged**